### ISSUE TITLE
How to pass a symbol instead of string in .kitchen.yml

### ISSUE THREAD

**Issue by lyrua** [test-kitchen/test-kitchen#556]
Hi, with the new chef client 12, I ran into issue with SSL, I'm tryng to disable the SSL check by adding ssl_verify_mode in the .kitchen.yml for testing purpose at the moment:

.kitchen.yml

.......
suites:
  -name: default
    provisioner:
      client_rb:
        ssl_verify_mode: :verify_none
    run_list:
      - recipe[tomcat]
......

Either I pass it with or without double quotes, it still writes it as a string:

client.rb
node_name "default-centos"
.......
ssl_verify_mode ":verify_none"

I removed the double quotes in client.rb and it ran successfully locally. Is there a way to pass it as a symbol in .kitchen.yml or is there a workaround?

**Comment by jf647** [comment]
This also prevents testing cookbooks that use the new audit mode in chef-client 12.1.x.

Because the client_rb hash values are transformed using .inspect:

https://github.com/test-kitchen/test-kitchen/blob/master/lib/kitchen/provisioner/chef_base.rb#182

It's impossible to end up with a line like this in TK's client.rb:

    audit_mode :enabled

I've tried all of these:

    :enabled
    !ruby/sym enabled
    ':enabled'

And in each case I get this in client.rb:

    audit_mode ":enabled"

**Comment by jtimberman** [comment]
Rather than populating the client.rb, I think we should have an audit_mode setting for the provisioner that will pass the audit mode specified as a string to the command-line. E.g.,

```yaml
provisioner:
  name: chef_zero
  audit_mode: audit_only
```

```
chef-client --audit-mode audit-only
```

```yaml
provisioner:
  name: chef_zero
  audit_mode: enabled
```

```
chef-client --audit-mode enabled
```

**Comment by tyler-ball** [comment]
I think we should update our YAML parsing to turn values that start with `:` into symbols when we parse it into Ruby

**Comment by fnichol** [comment]
I think we're in luck: there should be support for symbol values starting in d85961246ee5192db3c1fc337ceb4042b769dd57 and further worked in 5e3e8baf50944d8646986193d9670e0ae9bf38c4 which live currently in the [windows-guest-support](https://github.com/test-kitchen/test-kitchen/tree/windows-guest-support) branch (soon to be merged to master).

Can someone here take that branch out for a quick spin and see if we have success? Thanks!

**Comment by tyler-ball** [comment]
Debugging information:

https://gist.github.com/smford22/16c0164508160f3c4f79

**Comment by tyler-ball** [comment]
[SafeYaml](https://github.com/dtao/safe_yaml) keeps us from marshalling yaml into objects, but it also prevents creating symbols (a potential DDoS-like attack).

The plan right now is to keep that behavior and modify the provisioners to un-serialize symbol-like-objects (strings that start with `:`) into symbols when writing out configs.

**Comment by fnichol** [comment]
The "MVP" fix for Chef-related provisioners is in #652, just waiting on CI for a merge. If we haven't nailed this one down, we can reopen or file a new issue.

@jtimberman (or someone else), could you drop an issue around better supporting audit mode, should be easy enough once we know what it should look like.

Thanks all!
